hc_core/src/instance.rs

```diff
@@ -1,46 +1,146 @@
+    pub fn dispatch(&mut self, action: Action) -> ActionWrapper {
+        let wrapper = ActionWrapper::new(action);
+        self.action_channel
+            .send(wrapper.clone())
+            .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
+        wrapper
+    }
+
+    pub fn dispatch_and_wait(&mut self, action: Action) {
+        let wrapper = ActionWrapper::new(action);
+        let wrapper_clone = wrapper.clone();
+
+        let (sender, receiver) = channel::<bool>();
+        let closure = move |state: &State| {
+            if state.history.contains(&wrapper_clone) {
+                sender
+                    .send(true)
+                    .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
+                true
+            } else {
+                false
+            }
+        };
+
+        let observer = Observer {
+            sensor: Box::new(closure),
+            done: false,
+        };
+
+        self.observer_channel
+            .send(observer)
+            .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
+
+        self.action_channel
+            .send(wrapper)
+            .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
+
+        receiver
+            .recv()
+            .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
     }
 
-    pub fn pending_actions(&self) -> &VecDeque<Action> {
-        &self.pending_actions
+    pub fn dispatch_with_observer<F>(&mut self, action: Action, closure: F)
+    where
+        F: 'static + FnMut(&State) -> bool + Send,
+    {
+        let observer = Observer {
+            sensor: Box::new(closure),
+            done: false,
+        };
+
+        self.observer_channel
+            .send(observer)
+            .expect("observer channel to be open");
+        self.dispatch(action);
     }
 
-    pub fn consume_next_action(&mut self) -> Result<(), HolochainError> {
-        if !self.pending_actions.is_empty() {
-            let result = self.pending_actions.pop_front();
-            match result {
-                None => {
-                    return Err(HolochainError::ErrorGeneric(
-                        "nothing to consume".to_string(),
-                    ))
+    pub fn start_action_loop(&mut self) {
+        let (tx_action, rx_action) = channel::<ActionWrapper>();
+        let (tx_observer, rx_observer) = channel::<Observer>();
+        self.action_channel = tx_action.clone();
+        self.observer_channel = tx_observer.clone();
+
+        let state_mutex = self.state.clone();
+
+        thread::spawn(move || {
+            let mut state_observers: Vec<Box<Observer>> = Vec::new();
+
+            loop {
+                match rx_action.recv_timeout(Duration::from_millis(400)) {
+                    Ok(action_wrapper) => {
+                        // Mutate state:
+                        {
+                            let mut state = state_mutex.write().unwrap();
+                            *state = state.reduce(action_wrapper, &tx_action);
```

Comment:
I think @lucksus used a RWmutex to make sure that reads of the state from other threads are clean.

Reply:
Yes. The state is shared between this thread and the main thread/other threads. This is the only write but there may be several other reading threads. A read happening during a write would cause a problem. Several reads are ok. That's why a RWLock that distinguishes between reads and writes. A regular mutex would work as well. Not mutexing would not work. Wouldn't even compile in Rust.